extensions/amp-analytics/0.1/visibility-model.js

```diff
@@ -115,6 +154,68 @@ export class VisibilityModel {
 
     /** @private {time} milliseconds since epoch */
     this.lastVisibleUpdateTime_ = 0;
+
+    /** @private {boolean} */
+    this.waitToRefresh_ = false;
+
+    /** @private {?number} */
+    this.scheduleResetId_ = null;
+  }
+
+  /**
+   * Refresh counter on visible reset.
+   * TODO: Right now all state value are scoped state values that gets reset.
+   * We may need to add support to global state values, that never reset like globalTotalVisibleTime.
+   * Note: loadTimeVisibility is an exception.
+   * @private
+   */
+  refresh_() {
+    this.eventPromise_ = new Promise(resolve => {
+      this.eventResolver_ = resolve;
+    });
+    this.eventPromise_.then(() => {
+      this.onTriggerObservable_.fire();
+    });
+    this.waitToRefresh_ = false;
+    this.scheduleResetId_ = null;
+    this.everMatchedVisibility_ = false;
+    this.matchesVisibility_ = false;
+    this.continuousTime_ = 0;
+    this.maxContinuousVisibleTime_ = 0;
+    this.totalVisibleTime_ = 0;
+    this.firstVisibleTime_ = 0;
+    this.firstSeenTime_ = 0;
+    this.lastSeenTime_ = 0;
+    this.lastVisibleTime_ = 0;
+    this.minVisiblePercentage_ = 0;
+    this.maxVisiblePercentage_ = 0;
+    this.lastVisibleUpdateTime_ = 0;
+    this.waitToRefresh_ = false;
+  }
+
+  /**
+   * Function that visibilityManager can used to dispose model or reset model
+   */
+  disposeOrReset() {
+    // We may need a maxIntervalWindow to stop visible event even with reset true
+    if (!this.reset_) {
+      this.dispose();
+      return;
+    }
+    if (this.resetInterval_) {
+      const now = Date.now();
+      const interval = now - this.firstVisibleTime_;
+      const diff = Math.max(0, this.resetInterval_ - interval);
+      this.ready_ = false;
+      dev().assert(!this.scheduleResetId_, 'Should not reset twice');
+      this.scheduleResetId_ = setTimeout(() => {
+        this.refresh_();
+        this.setReady(true);
+      }, diff);
+    } else {
+      // Reset after element falls out of [minPercentage, maxPercentage]
```

Comment:
Nit: Is (minPercentage, maxPercentage] more correct?

Reply:
you're right! fixed